src/thor/expansion_action.cc

```diff
@@ -0,0 +1,144 @@
+  }
+}
+} // namespace
+
+namespace valhalla {
+namespace thor {
+
+std::string thor_worker_t::expansion(Api& request) {
+  // time this whole method and save that statistic
+  measure_scope_time(request, "thor_worker_t::expansion");
+
+  auto options = request.options();
+  auto exp_action = options.expansion_action();
+  // default generalization to ~ zoom level 15
+  float gen_factor = options.has_generalize() ? options.generalize() : 21.4f;
+
+  // default the expansion geojson so its easy to add to as we go
+  Document dom;
+  dom.SetObject();
+  FillDom(dom, exp_action);
+
+  // a lambda that the path algorithm can call to add stuff to the dom
+  // route and isochrone produce different GeoJSON properties
+  auto track_expansion = [&dom, &exp_action,
+                          &gen_factor](baldr::GraphReader& reader, baldr::GraphId edgeid,
+                                       const char* algorithm = nullptr, const char* status = nullptr,
+                                       const float duration = 0, const float distance = 0,
+                                       const float cost = 0) {
+    auto tile = reader.GetGraphTile(edgeid);
+    if (tile == nullptr) {
+      LOG_ERROR("thor_worker_t::expansion error, tile no longer available" +
+                std::to_string(edgeid.Tile_Base()));
+      return;
+    }
+    const auto* edge = tile->directededge(edgeid);
+    auto shape = tile->edgeinfo(edge).shape();
+
+    if (!edge->forward())
+      std::reverse(shape.begin(), shape.end());
+    Polyline2<PointLL>::Generalize(shape, gen_factor, {}, false);
+
+    // make the geom
+    auto& a = dom.GetAllocator();
+    auto* coords = GetValueByPointer(dom, "/features/0/geometry/coordinates");
+    coords->GetArray().PushBack(Value(kArrayType), a);
+    auto& linestring = (*coords)[coords->Size() - 1];
+    for (const auto& p : shape) {
+      linestring.GetArray().PushBack(Value(kArrayType), a);
+      auto point = linestring[linestring.Size() - 1].GetArray();
+      point.PushBack(p.first, a);
+      point.PushBack(p.second, a);
+    }
+
+    // make the properties
+    if (exp_action == Options::route) {
+      SetValueByPointer(dom, "/properties/algorithm", algorithm);
+      GetValueByPointer(dom, "/features/0/properties/edge_ids")
+          ->GetArray()
+          .PushBack(static_cast<uint64_t>(edgeid), a);
+      GetValueByPointer(dom, "/features/0/properties/statuses")
+          ->GetArray()
+          .PushBack(Value{}.SetString(status, a), a);
+
+    } else {
+      GetValueByPointer(dom, "/features/0/properties/durations")
+          ->GetArray()
+          .PushBack(Value{}.SetInt(static_cast<u_int64_t>(duration)), a);
+      GetValueByPointer(dom, "/features/0/properties/distances")
+          ->GetArray()
+          .PushBack(Value{}.SetInt(distance), a);
+      GetValueByPointer(dom, "/features/0/properties/costs")
+          ->GetArray()
+          .PushBack(Value{}.SetInt(cost), a);
+    }
```

Comment:
i see. yeah to me i feel like we could merge these together. statuses still exist in dijkstras but we could omit them for isochrones, and having the edge label info for all the other algorithms would be very helpful imho

Reply:
yeah, I'm just refactoring the whole thing (and some logic change), so this will look/feel better and be more consistent. I'll remove the draft when I'm ready for review:)